### ISSUE TITLE
Type definitions

### ISSUE THREAD

**Issue by matthieugomez** [spencerlyon2/CompEcon.jl#11]
Not an issue per se, but I have two questions about the way you define types:
- Why do you use immutable types? It sounds like immutable types are passed by values, which does not sound efficient. From the documentation:
  
  " An object with an immutable type is passed around (both in assignment statements and in function calls) by copying, whereas a mutable type is passed around by reference."
- I think the documentation advises to avoid abstract types in field definitions.
  http://julia.readthedocs.org/en/latest/manual/faq/#how-do-abstract-or-ambiguous-fields-in-types-interact-with-the-compiler.

**Comment by sglyon** [comment]
- I'm under the impression that using immutable types can only help make your code faster. I'll have to look into inefficiencies when passing them around.
- You are spot on with avoiding abstract types in field definitions. I still need to comb through the type system another time with two main goals in mind: 1. reduce reliance on abstract types and 2. reduce type instability (this is a _big_ performance killer and is quite prevalent in this code)

**Comment by sglyon** [comment]
Any suggestions for how to do this would be welcome

**Comment by matthieugomez** [comment]
About abstract type, you just need to add more parameters:

Instead of

``` julia
immutable Basis{N}
    basistype::Vector{BasisFamily}  # Basis family
    n::Vector{Int}                  # number of points and/or basis functions
    a::Vector{Float64}              # lower bound of domain
    b::Vector{Float64}              # upper bound of domain
    params::Vector{BasisParams}     # params to construct basis
end

immutable BasisStructure{BST<:ABSR}
    order::Matrix{Int}
    vals::Array{AbstractMatrix}
end
```

You can define

``` julia
immutable Basis{N, BF <: BasisFamily, BP <:BasisParams}
    basistype::Vector{BF}  # Basis family
    n::Vector{Int}                  # number of points and/or basis functions
    a::Vector{Float64}              # lower bound of domain
    b::Vector{Float64}              # upper bound of domain
    params::Vector{BP}     # params to construct basis
end

immutable BasisStructure{BST<:ABSR, V <: AbstractMatrix}
    order::Matrix{Int}
    vals::Array{V}
end
```

I don't understand what `N` and `BST` mean in your original definition so maybe I'm missing something

**Comment by sglyon** [comment]
Thanks for the comments. 

The reason I was stuck with abstract types instead of type parameters is because I wanted the flexibility to choose a different family of basis functions along different dimensions. Using type parameters would force all the basis to belong to the same family (`Cheb`, `Spline`, `Lin`) ect..

However, I'm finding that as I use the library I just stick with one family the whole time anyway.  I see two possible routes:
1. Move the current types/definitions to a new home and keep them as they are. Then users can still use them if they really want to. If we do this we should make clear that there is an expected performance penalty for doing this.
2. Just abandon the generality and enforce that a particular basis can have only one type of basis family for all dimensions. We lose a bit of generality, but probably gain more than enough performance for most potential users to be happy with the change

Thoughts?

---

The `N` is the dimensionality of the basis

The `BST` is actually crucial to how the heavy lifting actually gets done. See examples [here](https://github.com/spencerlyon2/CompEcon.jl/blob/c8c5e3f1fd759fdd12635dfe2fa25817843613a5/src/basis_structure.jl#L42-L104) and [here](https://github.com/spencerlyon2/CompEcon.jl/blob/c8c5e3f1fd759fdd12635dfe2fa25817843613a5/src/interp.jl#L105-L189)

**Comment by sglyon** [comment]
I should also mention that the `AbstractMatrix` is part of the `BasisStructure` definition in order to support multiple families in the same interpolation scheme. For the `Cheb` family `vals`  will be a standard `Array{Float64, 2}`, but for the `Spline` and `Lin` family the corresponding element of `vals`  will be `Base.SparseMatrix.SparseMatrixCSC`

**Comment by matthieugomez** [comment]
The types I defined still contains your type as a special case (i.e. it allows for the flexibility along dimensions). If all basis are similar in every dimension, then BF is `Cheb`. If basis are not similar, then BF is `BasisFamily`, and the corresponding type is equivalent to the type you're defining now.

**Comment by sglyon** [comment]
Yeah, I suppose that is true. 

It is a little funky to work with though:

```
julia> immutable Foo{T<:Real}
           a::T
           b::T
       end

julia> Foo(1.0, 2.0)
Foo{Float64}(1.0,2.0)

julia> Foo(1.0, 2)
ERROR: MethodError: `convert` has no method matching convert(::Type{Foo{T<:Real}}, ::Float64, ::Int64)
This may have arisen from a call to the constructor Foo{T<:Real}(...),
since type constructors fall back to convert methods.
Closest candidates are:
  Foo{T<:Real}(::T<:Real, ::T<:Real)
  call{T}(::Type{T}, ::Any)
  convert{T}(::Type{T}, ::T)
 in call at essentials.jl:57

julia>

julia> Foo{Real}(1.0, 2)
Foo{Real}(1.0,2)
```

I should add the need for specifying the abstract type parameter could easily be taken care of with specialized constructors...

**Comment by matthieugomez** [comment]
I don't have enough experience with parametric types (or interpolations to begin with!) to know if these situations are really problematic. That being said, this particular constructor issue is discussed in the documentation: http://julia.readthedocs.org/en/latest/manual/constructors/#parametric-constructors

**Comment by sglyon** [comment]
I think implementing your suggestions here would be a great idea. 

Are you up for working through a pull request?

If not I might try to tackle this later today once I finish working up #8

**Comment by matthieugomez** [comment]
No, sorry. I'd spend too much time just getting familiarized with the package - I don't even know the compecon package in Matlab is (I was just curious enough to read the README).

**Comment by sglyon** [comment]
No problem. 

Thanks for the helpful comments here. They solved some of the design questions I have had in the back of my head for a while now.

**Comment by matthieugomez** [comment]
Thank you! - I learned a lot using the quantecon package.

**Comment by sglyon** [comment]
That's great to hear! I was going to suggest checking it out, but you've already seen it!

**Comment by matthieugomez** [comment]
What about modifying your tutorial on types on QuantEcon to talk about it?
The tutorial constructs a type AR1 with abstract fields Real [here](http://quant-econ.net/jl/types_methods.html) - you only introduce type parameters as a way to ensure two fields have the same type.

A simple example

``` julia
abstract ProblemParameters

type ProblemParameters1 <:  ProblemParameters
    n::Integer
end

type ProblemParameters2{T} <: ProblemParameters
    n::T
end

function f(x::ProblemParameters)
    a = zero(Float64)
    for i in 1:1_000_000
        a += sqrt(x.n)
    end
end


@time f(ProblemParameters1(1))
# elapsed time: 0.074144595 seconds (32000096 bytes allocated, 43.78% gc time)
@time f(ProblemParameters2(1))
# elapsed time: 0.000307818 seconds (96 bytes allocated)
```

**Comment by sglyon** [comment]
Yes, that would be a great example to include there.

Thanks for the note. 

I'll open an issue in the (private) repo for the website source.

**Comment by sglyon** [comment]
Closed by be062591d0be561731497717403a749810dbe6c8

Thanks @matthieugomez !
